pkg/query-service/app/clickhouseReader/reader.go

```diff
@@ -1551,6 +1551,291 @@ func (r *ClickHouseReader) setTTLTraces(ctx context.Context, orgID string, param
+}
+
+func (r *ClickHouseReader) GetCustomRetentionTTL(ctx context.Context, orgID string) (*model.GetCustomRetentionTTLResponse, *model.ApiError) {
+	// Get the latest custom retention TTL setting
+	customTTL := new(types.TTLSetting)
+	err := r.sqlDB.BunDB().NewSelect().
+		Model(customTTL).
+		Where("org_id = ?", orgID).
+		Where("table_name = ?", r.logsDB+"."+r.logsLocalTableV2).
+		OrderExpr("created_at DESC").
+		Limit(1).
+		Scan(ctx)
+
+	if err != nil && err != sql.ErrNoRows {
+		zap.L().Error("Error in processing sql query", zap.Error(err))
+		return nil, &model.ApiError{Typ: model.ErrorExec, Err: fmt.Errorf("error in processing get custom retention ttl query")}
+	}
+
+	if err == sql.ErrNoRows {
+		return &model.GetCustomRetentionTTLResponse{
+			DefaultTTLDays: 15,
+			ResourceRules:  []model.CustomRetentionRule{},
+			Status:         constants.StatusFailed,
+		}, nil
+	}
+
+	var resourceRules []model.CustomRetentionRule
+	if customTTL.ResourceRules != "" {
+		if err := json.Unmarshal([]byte(customTTL.ResourceRules), &resourceRules); err != nil {
+			zap.L().Error("Error parsing resource rules", zap.Error(err))
+			resourceRules = []model.CustomRetentionRule{}
+		}
+	}
+
+	return &model.GetCustomRetentionTTLResponse{
+		DefaultTTLDays: customTTL.TTL,
+		ResourceRules:  resourceRules,
+		Status:         customTTL.Status,
+	}, nil
+}
+
+func (r *ClickHouseReader) checkCustomRetentionTTLStatusItem(ctx context.Context, orgID string, tableName string) (*types.TTLSetting, *model.ApiError) {
+	ttl := new(types.TTLSetting)
+	err := r.sqlDB.BunDB().NewSelect().
+		Model(ttl).
+		Where("table_name = ?", tableName).
+		Where("org_id = ?", orgID).
+		OrderExpr("created_at DESC").
+		Limit(1).
+		Scan(ctx)
+
+	if err != nil && err != sql.ErrNoRows {
+		zap.L().Error("Error in processing sql query", zap.Error(err))
+		return ttl, &model.ApiError{Typ: model.ErrorExec, Err: fmt.Errorf("error in processing custom_retention_ttl_status check sql query")}
+	}
+	return ttl, nil
+}
+
+func (r *ClickHouseReader) updateCustomRetentionTTLStatus(ctx context.Context, orgID, tableName, status string) {
+	statusItem, err := r.checkCustomRetentionTTLStatusItem(ctx, orgID, tableName)
+	if err == nil && statusItem != nil {
+		_, dbErr := r.sqlDB.BunDB().NewUpdate().
+			Model(new(types.TTLSetting)).
+			Set("updated_at = ?", time.Now()).
+			Set("status = ?", status).
+			Where("id = ?", statusItem.ID.StringValue()).
+			Exec(ctx)
+		if dbErr != nil {
+			zap.L().Error("Error in processing custom_retention_ttl_status update sql query", zap.Error(dbErr))
+		}
+	}
+}
+
+// Enhanced validation function to handle multiple conditions
+func (r *ClickHouseReader) validateResourceRules(ctx context.Context, resourceRules []model.CustomRetentionRule) *model.ApiError {
+	if len(resourceRules) == 0 {
+		return nil
+	}
+
+	// Collect all unique keys from all conditions
+	var allKeys []string
+	keySet := make(map[string]struct{})
+
+	for _, rule := range resourceRules {
```

Comment:
add more checks like for the same set of key values, multiple expressions cannot be added.